pkg/lib/_internal/packagemanager-abstract.ts

```diff
@@ -0,0 +1,99 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2025 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <https://www.gnu.org/licenses/>.
+ */
+
+export interface ProgressData {
+    // PackageManager waits on a lock or another operation to finish
+    waiting: boolean
+    // If not null, the operation can be cancelled
+    cancel?: (() => void) | null
+    // FIXME, mistake? should be percentage?
+    absolute_percentage: number
+}
+
+export type ProgressCB = (data: ProgressData) => void;
+export interface InstallProgressData extends ProgressData {
+    // One of the Enum states
+    info: number
+    // Package being downloaded or installed
+    package: string
+}
+
+export type InstallProgressCB = (data: InstallProgressData) => void;
+
+export interface MissingPackages {
+    // Packages that were requested, are currently not installed, and can be installed
+    missing_names: string[]
+    // The full package IDs corresponding to missing_names (the ID different per backend)
+    missing_ids: number[]
+    // Packages that were requested, are currently not installed, but can't be found in any repository
+    unavailable_names: string[]
+    // The to be installed package IDs, temporary kept (the ID different per backend)
+    install_ids: number[]
+    // The full package IDs corresponding to remove_names (the ID different per backend)
+    remove_ids: number[]
+
+    // If unavailable_names is empty, a simulated installation of the missing packages
+    // is done and the result also contains these fields:
```

Comment:
Does dnfdaemon support simulated install?

Reply:
Maybe... what I use is:

```
            await call(session, "org.rpm.dnf.v0.rpm.Rpm", "install", [pkgnames, {}]);
            const [transaction_items, result] = await call(session, "org.rpm.dnf.v0.Goal", "resolve", [{}]) as InstallResolveResult;
```

We would need more specific tests honestly to hit remove_names, extra_names is touched by pcp as far as I know.

Reply:
tests should install a package "canary", and the fake "pcp" package then gets `Conflicts: canary`, this should trigger a removal.

Reply:
Agreed

Reply:
This was done in `check-packagekit` in the new install dialog tests.